### ISSUE TITLE
mlpack KMeans class much slower than armadillo kmeans()

### ISSUE THREAD

**Issue by jciminogh** [mlpack/mlpack#514]
I seem to be getting much faster clustering from armadillo's [kmeans()](http://arma.sourceforge.net/docs.html#kmeans) function in comparison to mlpack's kmeans::KMeans<> class, which is about 2x to 6x slower. Using the latest mlpack code from the git repo. Am I doing something wrong?

Using the code below I get the following timings on my machine (Intel i5, 64 bit, g++ version 5.3).

Compiled without openmp:
`g++ kmeans_test.cpp -o kmeans_test -O3 -std=c++11 -larmadillo -lmlpack`
mlpack_kmeans time: 17.3024
arma::kmeans time: 9.16399

Compiled with openmp:
`g++ kmeans_test.cpp -o kmeans_test -O3 -std=c++11 -larmadillo -lmlpack -fopenmp`
mlpack_kmeans time: 17.7575
arma::kmeans time: 2.87675

```
#include <fstream>
#include <mlpack/methods/kmeans/kmeans.hpp>
#include <armadillo>

int main() {
    arma::uword dims = 20; // number of dimensions
    arma::uword samples = 5000000;

    arma::uword max_iterations = 10;
    arma::uword k = 10;  // number of clusters

    arma::arma_rng::set_seed_random(); // random start

    std::cout << "Generating some synthetic data ..." << std::endl;

    arma::mat data(dims, samples, arma::fill::zeros);

    // generate data with unique centroids, added with a small amount of noise
    for (arma::uword i=0; i<samples; i++) {
        arma::uword c = as_scalar( arma::randi<arma::uvec>(1, arma::distr_param(0,k-1)) );
        data.col(i) = arma::linspace<arma::vec>(c, c+dims-1, dims) + 0.25*arma::randn<arma::vec>(dims);
      }

    arma::wall_clock timer;

    std::cout << "mlpack_kmeans start ..." << std::endl;

    arma::Row<size_t> mlpack_assignments;
    arma::mat mlpack_centroids;

    mlpack::kmeans::KMeans<> mlpack_kmeans(max_iterations);

    timer.tic();
    mlpack_kmeans.Cluster(data, k, mlpack_assignments, mlpack_centroids);

    std::cout << "mlpack_kmeans time: " << timer.toc() << std::endl;

    std::cout << "---" << std::endl;
    std::cout << "arma::kmeans start ..." << std::endl;

    arma::mat arma_centroids;

    timer.tic();
    arma::kmeans(arma_centroids, data, k, arma::random_subset, max_iterations, false);

    std::cout << "arma::kmeans time: " << timer.toc() << std::endl;

    std::cout << "---" << std::endl;

    mlpack_centroids.print("mlpack_centroids:");
    arma_centroids.print("arma_centroids:");

    return 0;
}
```

**Comment by rcurtin** [comment]
mlpack's k-means does not use OpenMP at this point in time.  So yes, what you're seeing is normal.  But do keep in mind that mlpack's k-means implementation is far more flexible, allowing you to use five or six different algorithms---one of which will probably be faster than the default O(kN) algorithm.  (Try it;  for your dataset I'd expect Hamerly's algorithm or the single-tree Pelleg-Moore algorithm to be fastest.)

**Comment by rcurtin** [comment]
I kept thinking about this, and I realized this is an unfair comparison, because if you aren't starting from the exact same starting centroids, the algorithm may converge in different ways.  If you specify random starting points and use `arma::keep_existing`, that will be a fairer comparison.  I also somehow missed the "without openmp" numbers; sorry about that.  It's possible that all the abstractions used for mlpack's k-means algorithm cause some amount of slowdown.  It should be avoidable, but it may be a while until I get around to improving it (if you feel so inclined, I'm happy to help with the contribution!).

**Comment by jciminogh** [comment]
I have tried many starting points like random and same centroids for mlpack and arma. So far the armadillo kmeans() function is typically about 2x faster. With openmp enabled it is always many times faster.

For my purposes the default clustering method is good so far, but I will try your suggestion with mlpack's other clustering methods.

**Comment by rcurtin** [comment]
Thanks for doing this extra testing.  I'm going to reopen this bug then, since it seems like there is good reason to believe that there actually is an efficiency issue (and also now that I'm more awake than when I originally responded to this).  The things that should happen are

 - improvement of the current `NaiveKMeans` code to at least be roughly equivalent to the Armadillo code, which may mean using the internal Armadillo code, depending
 - OpenMP at least for the `NaiveKMeans` strategy, possibly also for `ElkanKMeans` and `HamerlyKMeans`; the tree-based ones are a lot harder so OpenMP may not be as useful there...
 - production of some nice benchmarks and maybe addition of k-means to the benchmarking system (which unfortunately is down at the moment, but with any luck might come back up next week), in order to verify that mlpack's naive k-means implementation is at least as good as Armadillo's.

Anyway, it may be a while until I'm able to look into this more fully, but I will, when I have a chance...

**Comment by stereomatchingkiss** [comment]
Do you have any plan to implement mini-batch kmeans?

**Comment by rcurtin** [comment]
@stereomatchingkiss: not at the moment, but if you wanted to write support for that, please feel free.  We would have to figure out how to make that work with the existing `KMeans<>` class.

**Comment by mitsoschelsea** [comment]
Why don't you use pthreads to solve this? After several tests to bitonic shorting algorithm, pthreads was much faster than OpenMP. Approximately 5 to 6 times faster.

**Comment by Fractally** [comment]
I may be wrong, but from the code in the first post it appears as if mlpack also provides the assignments while armadillo does not. Could this not be a reason for some performance loss?

**Comment by rcurtin** [comment]
If so, that would be really slight, because the k-means algorithm requires you to find the closest cluster to every point already anyway.  You could get some RAM savings by not caching those assignments, but actual noticeable significant speedup I kind of doubt.

**Comment by cypro666** [comment]
Infact, dual-tree or pelleg-moore(same as x-means) are really fast enough!!!
The implement of naive-kmean now is not suitable for parallel using omp, it's “textbook style”， but also good for small dataset ：）
**I changed the details of NaiveKMeans::Iterate method directly to use arma::kmeans**

```cpp
template<typename MetricType, typename MatType>
double NaiveKMeans<MetricType, MatType>::Iterate(const arma::mat& centroids, arma::mat& newCentroids, arma::Col<size_t>& counts) {
    counts.zeros(centroids.n_cols); // never used, in fact
    newCentroids = centroids;

    arma::kmeans(newCentroids, dataset, centroids.n_cols, arma::keep_existing, 10, false);
    Log::Assert(newCentroids.n_cols == centroids.n_cols);

    // Now normalize the centroid.
    distanceCalculations += centroids.n_cols * dataset.n_cols;

    // Calculate cluster distortion for this iteration.
    double cNorm = 0.0;
    for (size_t i = 0; i < centroids.n_cols; ++i) {
        cNorm += std::pow(metric.Evaluate(centroids.col(i), newCentroids.col(i)), 2.0);
    }
    distanceCalculations += centroids.n_cols;

    return std::sqrt(cNorm);
}
```

Now it's as faster as armadillo's
But this should not be a correct, good or final solution *_*

**Comment by mentekid** [comment]
Closed with #1024
